agent/structs/config_entry_discoverychain.go

```diff
@@ -68,78 +66,70 @@ func (e *ServiceRouterConfigEntry) Validate() error {
 	// catch-all is configured for you, but at that point maybe you should just
 	// delete it so it will default?
 
 	for i, route := range e.Routes {
-		if route.Match == nil || route.Match.HTTP == nil {
-			continue
-		}
-
-		pathParts := 0
-		if route.Match.HTTP.PathExact != "" {
-			pathParts++
-		}
-		if route.Match.HTTP.PathPrefix != "" {
-			pathParts++
-		}
-		if route.Match.HTTP.PathRegex != "" {
-			pathParts++
-		}
-		if pathParts > 1 {
-			return fmt.Errorf("Route[%d] should only contain at most one of PathExact, PathPrefix, or PathRegex", i)
-		}
-
-		// TODO(rb): do some validation of PathExact and PathPrefix
-
-		for j, hdr := range route.Match.HTTP.Header {
-			if hdr.Name == "" {
-				return fmt.Errorf("Route[%d] Header[%d] missing required Name field", i, j)
-			}
-			hdrParts := 0
-			if hdr.Present {
-				hdrParts++
-			}
-			if hdr.Exact != "" {
-				hdrParts++
-			}
-			if hdr.Regex != "" {
-				hdrParts++
+		eligibleForPrefixRewrite := false
+		if route.Match != nil && route.Match.HTTP != nil {
+			pathParts := 0
+			if route.Match.HTTP.PathExact != "" {
+				eligibleForPrefixRewrite = true
+				pathParts++
+				if !strings.HasPrefix(route.Match.HTTP.PathExact, "/") {
+					return fmt.Errorf("Route[%d] PathExact doesn't start with '/': %q", i, route.Match.HTTP.PathExact)
+				}
 			}
-			if hdr.Prefix != "" {
-				hdrParts++
+			if route.Match.HTTP.PathPrefix != "" {
+				eligibleForPrefixRewrite = true
+				pathParts++
+				if !strings.HasPrefix(route.Match.HTTP.PathPrefix, "/") {
+					return fmt.Errorf("Route[%d] PathPrefix doesn't start with '/': %q", i, route.Match.HTTP.PathPrefix)
+				}
 			}
-			if hdr.Suffix != "" {
-				hdrParts++
+			if route.Match.HTTP.PathRegex != "" {
+				pathParts++
 			}
-			// "absent" is the bare invert=true
-			if (hdrParts == 0 && !hdr.Invert) || (hdrParts > 1) {
-				return fmt.Errorf("Route[%d] Header[%d] should only contain one of Present, Exact, Prefix, Suffix, or Regex (or just Invert)", i, j)
+			if pathParts > 1 {
+				return fmt.Errorf("Route[%d] should only contain at most one of PathExact, PathPrefix, or PathRegex", i)
 			}
-		}
 
-		for j, qm := range route.Match.HTTP.QueryParam {
-			if qm.Name == "" {
-				return fmt.Errorf("Route[%d] QueryParam[%d] missing required Name field", i, j)
+			for j, hdr := range route.Match.HTTP.Header {
+				if hdr.Name == "" {
+					return fmt.Errorf("Route[%d] Header[%d] missing required Name field", i, j)
+				}
+				hdrParts := 0
+				if hdr.Present {
+					hdrParts++
+				}
+				if hdr.Exact != "" {
+					hdrParts++
+				}
+				if hdr.Regex != "" {
+					hdrParts++
+				}
+				if hdr.Prefix != "" {
+					hdrParts++
+				}
+				if hdr.Suffix != "" {
+					hdrParts++
+				}
+				// "absent" is the bare invert=true
+				if (hdrParts == 0 && !hdr.Invert) || (hdrParts > 1) {
+					return fmt.Errorf("Route[%d] Header[%d] should only contain one of Present, Exact, Prefix, Suffix, or Regex (or just Invert)", i, j)
```

Comment:
I like this clear error but I'm not sure I understand `(or just Invert)`. It sounds like you are saying `Invert` is mutually exclusive with the rest too but why not just add it to the list in that case?

Reply:
I got myself all confused here when validating how to say "header is absent". I've corrected it. Previously I was mistakenly taking `{Invert: true}` to mean `{Present:false}` which makes no sense. The correction is to require you to say the saner thing of `{Present: true, Invert: true}` to express absent.